### PR TITLE
UPS Operator integration

### PR DESCRIPTION
**Pull request by ziccardi**
## Motivation
https://issues.jboss.org/browse/AEROGEAR-9435

## What
- Added ability to customise the resource management for each of the resource type
- Added the PushVariantCR object
- Added the PushApplicationCR object
- Customised the behaviour of the PushVariantCR Manager so that variants are managed

## How to test

1. Run the MDC as usual
2. Install the UPS operator in the `mobile-console` namespace
3. Use the UPS operator to create a `UnifiedPush` instance
4. Create an app in MDC
5. Click to the app details
6. Click on mobile services
7. Try to bind and unbind the Push Service. Note that Push Service can be bound 2 times, one for android and one for iOS

## Tasks:
- [x] Add ability to bind an android variant
- [x] Add ability to bind an ios variant 
- [x] Add ability to unbind the variants
- [x] Customise the watchers to be able to support services using more than one custom resource
- [x] Add ability to automatically create the UPS application when creating a variant if it doesn't already exists

## KNOWN BUGS

There is a bug in the binding window: if you select ios and then try to move back to android, the fields does not gets updated. This will be fixed as part of another issue.

### CODE REVIEW

File: server/server.js [hunk 1]
```diff
@@ -158,7 +159,10 @@ async function initKubeClient() {
     const conf =
       process.env.NODE_ENV === 'production' ? Request.config.getInCluster() : Request.config.fromKubeconfig();
     const backend = new Request(conf);
-    const kubeclient = new Client({ backend });
+    backend.requestOptions.insecureSkipTlsVerify = true;
+    backend.requestOptions.strictSSL = false;
```
Comment (ziccardi) [inline server/server.js]: these are just to make MDC work with a remote openshift using selfsigned certs. Will be removed before merging the PR.

File: src/services/crmanagers/PushVariantResourceManager.js [hunk 2]
```diff
@@ -0,0 +1,84 @@
+import _ from 'lodash-es';
+import { GenericResourceManager } from './GenericResourceManager';
+import { PushApplicationCR } from '../../models/mobileservices/pushapplicationcr';
+
+function poll(fnToBePolled, checkOutcome, timeout, interval) {
```
Comment (craicoverflow) [inline src/services/crmanagers/PushVariantResourceManager.js]:
```suggestion
function poll(fnToBePolled, checkOutcome, timeout, interval = 100) {
```

**Comment by wei-lee** [comment]
@ziccardi do you have a cluster that I can quickly verify the changes?

**Comment by ziccardi** [comment]
@wei-lee I used https://master.batman-e909.openshiftworkshop.com. You can easily edit your `development.sh` script to point there (just change the `minishift` ip and port). If it complains for the self-signed certificate just change the server/server.js as follows:
```diff
diff --git a/server/server.js b/server/server.js
index c609a6fd..d67c7f7d 100644
--- a/server/server.js
+++ b/server/server.js
@@ -159,7 +159,10 @@ async function initKubeClient() {
     const conf =
       process.env.NODE_ENV === 'production' ? Request.config.getInCluster() : Request.config.fromKubeconfig();
     const backend = new Request(conf);
-    const kubeclient = new Client({ backend });
+    backend.requestOptions.insecureSkipTlsVerify = true;
+    backend.requestOptions.strictSSL = false;
+
+    const kubeclient = new Client({ backend, config: { insecureSkipTlsVerify: true } });
     await kubeclient.loadSpec();
     kubeclient.addCustomResourceDefinition(mobileClientCRD);
     kubeclient.addCustomResourceDefinition(mobileSecurityServiceCRD);

```

**Comment by jhellar** [comment]
Followed the verification steps - works for me. Noticed just 2 minor issues:

- in binding details `UPS Admin Console URL` and `Push Application` urls are not correct
- when clicking on the three dots button for binding, when trying to delete it, there is `Delete AndroidVariant` - missing space before `Variant` (same for iOS)

**Comment by ziccardi** [comment]
@wei-lee Added an `INSECURE_SERVER` option to allow untrusted/selsigned certificates on remote openshift
@jhellar 👍 Looking into that now

**Comment by ziccardi** [comment]
@jhellar 

> in binding details UPS Admin Console URL and Push Application urls are not correct

That's because the installer has not been run. When it will be run, they will be correct.

> when clicking on the three dots button for binding, when trying to delete it, there is Delete AndroidVariant - missing space before Variant (same for iOS)

We will fix it as part of another issue

**Pull request merged**